articles/121_realtime_proposal.md

```diff
@@ -0,0 +1,227 @@
+
+## Requirements and Implementation of Real-Time Systems
+
+### System requirements
+
+The requirements of a real-time system vary depending on the use case.
+At its core, the real-time requirement of a system has two components:
+
+- Latency
+
+  - Update period
+  - Predictability
+
+- Failure mode
+
+  - How to react to a missed deadline
+
+Language referring to hard/soft/firm real-time systems generally refers to the failure mode.
+A hard real-time system treats a missed deadline as a system failure.
+A soft real-time system tries to meet deadlines but does not fail when a deadline is missed.
+A firm real-time system discards computations made by missed deadlines and may degrade its performance requirement in order to adapt to a missed deadline.
+
+Though in this document we consider the failure mode orthogonal to latency characteristics, a hard real-time failure mode is often associated with high predictability (low jitter).
+Safety-critical systems often require a hard real-time with system with high predictability.
+
+### Implementation choices: computer architecture and OS
+
+Processors vary in capability, number of cores, memory size, and instruction set architectures, some which are better suited to real-time performance than others.
+The selection of computer architecture may be due to cost, size, or energy efficiency, which could set a constraint for the rest of the system.
+A single robot may contain multiple computers with a variety of processor types, from the main control board to sensors.
+
+Another major consideration when designing a real-time system is the choice of operating system.
+A true "hard" real-time operating system is truly deterministic.
+"Soft" real-time operating systems are designed to meet most deadlines with low jitter, but are nondeterministic.
+Most conventional user-facing operating systems are not designed with fairness and predictability as the first priorities, often because they have to support graphical displays and interactivity.
+Here are some features critical to predictable performance that are often exposed to the user in an RTOS, but not in a conventional OS (see Wikipedia, [Real-time operating system](https://en.wikipedia.org/wiki/Real-time_operating_system)).
+For more information about why these features matter, refer to the [introductory article on real-time systems](http://design.ros2.org/articles/realtime_background.html).
+
+- Task scheduling
+
+  - Expose thread priority and deterministic schedulers that incorporate priority
+  - Allow full thread preemption
+
+- Concurrency and synchronization primitives
+
+  - Floored semaphores (allow a higher priority thread to take control of locked semaphore)
+
+- Interrupt handling
+
+  - Disable interrupts
+  - Mask interrupts (temporarily suspend interrupts in a critical section)
+
+- Memory allocation
+
+Some real-time systems have a non real-time component, e.g. for graphical display or other user interaction.
```

Comment:
Or for communication with non-real-time parts of a robot's software. ROS1 and Orocos being used together are the perfect example here.